### PR TITLE
Add auto-yasnippet

### PR DESCRIPTION
**Pull request by tuhdo**
Make it easy to create snippets. For example:

- Sample code:

```C++
const Point<3> curl(grad[~2][~1] - grad[~1][~2],
```

- Run `aya-create` on the selected region above.
- See the `~` removed.
- Run `aya-expand` to insert the snippet again, except this time we will be able to move through the positions masked with `~` earlier using `TAB` or `S-TAB`. It is useful when combined with a macro to increase the counter. The nice thing is, the one marked with the same index automatically change accordingly as you type.
- To save the snippet, run `aya-persist-snippet`. A prompt asks for snippet name, and the positions with `~` earlier is replaced with appropriate variables.

### CODE REVIEW

**Comment by jasoncyu** [comment]
nice! what do you think of adding comments giving a mnemonic hint? like

```
     ;; aya *y*ank
     "iy" 'aya-create
     ;; aya *e*xpand
     "ie" 'aya-expand
     ;; aya *w*rite
     "iw" 'aya-persist-snippet)))
```

i saw it in the python layer and thought it was helpful

**Comment by tuhdo** [comment]
Yeah i should update the doc as well.

**Comment by tuhdo** [comment]
Updated it.

File: contrib/auto-completion/README.md [hunk 1]
```diff
@@ -9,7 +9,7 @@
     - [Configuration](#configuration)
         - [Key bindings](#key-bindings)
         - [Tooltips](#tooltips)
-        - [Sort results by usage](#sort-results-by-usage)
+n        - [Sort results by usage](#sort-results-by-usage)
```
Comment (person808) [inline contrib/auto-completion/README.md]: You accidentally a letter here

**Comment by tuhdo** [comment]
Fixed.

**Comment by syl20bnr** [comment]
Thank you ! Cool addition to the snippets and macros.
I changed a bit the key bindings to take only one key in the `i` space and made an auto-switch to insert state when expanding for convenience, related commit: https://github.com/syl20bnr/spacemacs/commit/5ee7203
Cherry-picked into `develop`, you can safely delete your branch.

**Comment by tuhdo** [comment]
wow that's 5 keys to press for Vim style and 6 for Emacs. I think you should revert back the keys to what they were, since `auto-yasnippet` is meant to be used frequently in during programming session, since it allows you to create temporary snippet to quickly insert duplicate things but not necessary always save. You should [read this article](http://oremacs.com/2015/04/30/gcc-macros-auto-yasnippet/) for one such example.

**Comment by syl20bnr** [comment]
This is a niche package that very few users will really use. Big users of auto-yasnippet will probably put the bindings on a modifier or behind `o`.

A better package to do this stuff is `multiple-cursors` where you can paste the same text and only when you pasted all the text you put your cursors and make the modifications. `auto-yasnippet` workflow is not very efficient because after each `aya-expand` you have to fill the gaps which make it less fluent. For instance my first modification to your PR was to make a micro-state... it was useless since you can `aya-expand` and only after then fill the gaps.

**Comment by tuhdo** [comment]
I don't think `multiple-cursors` can do [this example](https://github.com/abo-abo/auto-yasnippet#c). Or, create a `LightOff` class from the original LightOn quickly: [example](https://github.com/abo-abo/auto-yasnippet#java). So, instead of yanking existing class, paste it, manually modifying each occurence of `On` to `Off`, you only need to modify a single `On` to `Off` and  the rest follow, without having to narrow or whatever extra action. You can think of this like a special case of copy/paste.

Aslo, `multiple-curosrs` won't also save your snippets permanently and automatically inserting snippet variables for you.

**Comment by syl20bnr** [comment]
The example can be done with MC and the workflow is better in my opinion.
You start with:
```
const Point<3> curl(grad[x][y] - grad[y][x],
                    grad[x][y] - grad[y][x],
                    grad[x][y] - grad[y][x]);
```
Then put 2 cursors on the `x` and edit, you repeat for the `y`. This is a better separation of concerns, first I write the text, then I modify it, the two types of actions are not intertwined.

For the On, Off example, this is total bullshit, there is absolutely no improvements with a regular workflow where you would copy/paste the `On` classe as is and then edit the `On` with `iedit`. The workflow is just reversed, with `auto-yasnippet` you first think about where you have to modify the pasted text, in a regular workflow you think about it aftwerard, in both cases you can make the _same_ mistakes and forgot some occurrences to replace.

The to save a snippet may be useful but not so much........ creating a snippet is one thing, but what matters is to not forgot it and effectively reuse it. I doubt that snippet created on the fly are good candidates for good snippets with those properties.

**Comment by tuhdo** [comment]
@geo7 this PR is merged. You should have a look at it. I think you will really like it how easy you can create and save a snippet from your original text with minimal effort. Basically, where you want your snippet varied in the original text, just insert a `~` character before it. After you cal `aya-create` (currently bound to `SPC i S c`), it will remove all the `~` characters you inserted earlier and save the snippet for insertion with `aya-expand` (bound to `SPC i S e`). You can save the snippet with `aya-persist-snippet` (bound to `SPC i S w`).

**Comment by tuhdo** [comment]
> For the On, Off example, this is total bullshit, there is absolutely no improvements with a regular workflow where you would copy/paste the On classe as is and then edit the On with iedit or something else. The workflow is just reversed, 

No, it's not. If you use `iedit`, it modifies everything in the buffer including your original text and you have to narrow to the newly pasted text, then widen it after that.

So, what if the example `grad[x][y] - grad[y][x]` does not require insert `x` or `y` but instead variables? i.e.

```
const Point<3> curl(grad[a][b] - grad[b][a],
                    grad[c][d] - grad[d][c],
                    grad[e][f] - grad[f][e]);
```

> To save a snippet may be useful but not so much........ creating a snippet is one thing, but what matters is to not forgot it and effectively reuse it. I doubt that snippet created on the fly are good candidates for good snippets with those properties.

Have you tried saving the snippets? It prompts you a name, and open a new snippet buffer with snippet text inserted with appropriate variables. Let's wait for @geo7, the one who desperately needs to have a good snippet system to make a snippet from Vim.

**Pull request closed without merging**